src/main/groovy/shiro/ShiroGrailsPlugin.groovy

```diff
@@ -133,7 +135,15 @@ Enables Grails applications to take advantage of the Apache Shiro security layer
             }
 
             // Default remember-me manager.
-            shiroRememberMeManager(CookieRememberMeManager)
+            def cipherKeyFromConfig = grailsApplication.config.getProperty('security.shiro.rememberMe.cipherKey')
+            if (cipherKeyFromConfig) {
+                if (!(cipherKeyFromConfig.length() * 8 in [128, 192, 256])) {
```

Comment:
So, the reason I want to know the type, and check for it, is what if the key is a string and therefore utf16? Are we expecting a byte array? should we enforce 8 bit ASCII encoded key?

Reply:
I'd like to see a test for this if possible? I can help if you like

Reply:
I am using string as the key which is then converted to byte array using getBytes with utf-8 encoding to get consistent length.

Reply:
utf8 length is variable, it supports backwards compatibility to ASCII, but it's a multi-byte character format. So, there is the potential if someone uses a unicode char in a string that has 2 or three bytes that when you get the bytes and check the bit length that they will get an error that won't be easy to understand - now thats the problem, how to fix it?

Perhaps read in, force to bytes, truncate at 16, 24, 32 bytes and just use that. Throw an error if short, and log a warning if too long?

Reply:
Also a script to generate a key will prevent most problems here :-)

Reply:
'getProperty('security.shiro.rememberMe.cipherKey', String, null)

Reply:
e.g. try this in a groovy console
  String key = "ĠġĢģĤĥĦħĨĩĪīĬsss"
  assert key.size() == 16
  byte[] keyBytes = key.getBytes("UTF-8")
  assert keyBytes.size() == 16

Reply:
yup, this is awkward, will think about it.

Reply:
Not sure if truncating the byte array is a good solution. You are left with a key which is not used as a whole but a weird part of it (it can cutoff a 2/3byte character). Maybe we can throw the warning but i would be much more satisfied if we can somehow do it to protect the user from confusion and mistakes.

There is always the possibility to require a byte array of the specified lenghts (16,24,32) in the config, but that is rather inconvenient.

Reply:
So base64 encoded srtings - as I suggested earlier, generated by a script would solve it - I'll pull and merge this and add a script you can try out...